File: android/app/src/main/java/com/gamebooster/app/device/HardwareDisplayController.java
package com.gamebooster.app.device;

import android.app.Activity;
import android.content.Context;
import android.hardware.display.DisplayManager;
import android.os.Build;
import android.provider.Settings;
import android.util.Log;
import android.view.Display;
import android.view.Surface;
import android.view.WindowManager;

import java.lang.reflect.Method;
import java.util.ArrayList;

import com.gamebooster.app.config.GameProfileAutoConfigurator;

/**
 * Reads Android-reported display modes and applies display preferences only
 * where the user has granted Modify system settings. It never claims to change
 * a separate game's frame cap or to create a mode the panel does not expose.
 */
public final class HardwareDisplayController {

    private static final String TAG = "HardwareDisplayCtrl";

    private HardwareDisplayController() {}

    public static float getMaxHardwareRefreshRate(Context context) {
        Display.Mode mode = getMaxHardwareMode(context);
        return mode != null ? mode.getRefreshRate() : 0f;
    }

    public static Display.Mode getMaxHardwareMode(Context context) {
        if (context == null) return null;
        DisplayManager manager = (DisplayManager) context.getSystemService(Context.DISPLAY_SERVICE);
        Display display = manager != null ? manager.getDisplay(Display.DEFAULT_DISPLAY) : null;
        if (display == null) return null;

        Display.Mode best = null;
        for (Display.Mode mode : display.getSupportedModes()) {
            if (mode != null && (best == null || mode.getRefreshRate() > best.getRefreshRate())) {
                best = mode;
            }
        }
        return best;
    }

    /** Returns a physical display mode reported by Android, or zero if unavailable. */
    public static int resolveSupportedRefreshRate(Context context, int requestedHz) {
        if (context == null) return 0;
        DisplayManager manager = (DisplayManager) context.getSystemService(Context.DISPLAY_SERVICE);
        Display display = manager != null ? manager.getDisplay(Display.DEFAULT_DISPLAY) : null;
        if (display == null) return 0;

        ArrayList<Integer> rates = new ArrayList<>();
        for (Display.Mode mode : display.getSupportedModes()) {
            if (mode != null) rates.add(Math.round(mode.getRefreshRate()));
        }
        return RefreshRatePolicy.resolveRate(rates, requestedHz);
    }

    /** Applies the highest physical display mode to this app's own window. */
    public static void applyMaxRefreshRateToWindow(Activity activity) {
        if (activity == null) return;
        try {
            Display.Mode maxMode = getMaxHardwareMode(activity);
            if (maxMode == null) return;
            WindowManager.LayoutParams parameters = activity.getWindow().getAttributes();
            parameters.preferredDisplayModeId = maxMode.getModeId();
            activity.getWindow().setAttributes(parameters);
            Log.i(TAG, "Launcher window prefers " + maxMode.getRefreshRate() + "Hz");
        } catch (Throwable t) {
            Log.w(TAG, "Unable to set launcher display preference", t);
        }
    }

    /** Applies a frame-rate hint to a Surface owned by this app. */
    public static void applySurfaceFrameRate(Surface surface, float targetFps) {
        if (surface == null || targetFps <= 0) return;
        try {
            surface.setFrameRate(targetFps, Surface.FRAME_RATE_COMPATIBILITY_FIXED_SOURCE,
                    Surface.CHANGE_FRAME_RATE_ONLY_IF_SEAMLESS);
        } catch (Throwable t) {
            Log.d(TAG, "Surface frame-rate hint unavailable: " + t.getMessage());
        }

        if (Build.VERSION.SDK_INT >= 34) {
            try {
                Method method = Surface.class.getMethod("setFrameRateCategory", int.class);
                method.invoke(surface, 3 /* FRAME_RATE_CATEGORY_HIGH */);
            } catch (Throwable ignored) {
                // Category hints are optional and device-dependent.
            }
        }
    }

    /**
     * Enforces the target refresh rate via the Settings.System API AND directly via Shizuku/Root.
     *
     * Unlike the old implementation this method does NOT silently return false when
     * MODIFY_SYSTEM_SETTINGS is missing — it fires privileged shell commands as a fallback,
     * ensuring peak_refresh_rate + min_refresh_rate are NEVER left at 60Hz.
     *
     * @param requestedHz Target Hz (120 / 144 / 165 / 185).
     * @return true if at least one write path succeeded.
     */
    public static boolean forceUnlockSystemRefreshRate(Context context, int requestedHz) {
        if (context == null || requestedHz <= 0) return false;

        int targetHz = GameProfileAutoConfigurator.clampTargetFpsToDisplay(context, requestedHz);
        float targetF = (float) targetHz;
        boolean anySuccess = false;

        // ── Path A: Settings.System (public API) ─────────────────────────────────────
        if (Settings.System.canWrite(context)) {
            try {
                boolean peak = Settings.System.putFloat(context.getContentResolver(), "peak_refresh_rate", targetF);
                boolean min  = Settings.System.putFloat(context.getContentResolver(), "min_refresh_rate",  targetF);
                // Disable adaptive / match-content frame-rate — these silently drop Hz to 60
                Settings.System.putInt(context.getContentResolver(), "match_content_frame_rate", 0);
                Settings.Global.putInt(context.getContentResolver(), "match_content_frame_rate", 0);
                Log.i(TAG, "Settings.System enforced " + targetHz + "Hz (peak=" + peak + ", min=" + min + ")");
                anySuccess = peak || min;
            } catch (SecurityException e) {
                Log.w(TAG, "Settings.System write denied: " + e.getMessage());
            }
        } else {
            Log.i(TAG, "MODIFY_SYSTEM_SETTINGS not granted — escalating to privileged shell");
        }

        // ── Path B: Privileged shell via Shizuku/Root (fires regardless of canWrite) ─
        // These bypass the Android Settings permission gate entirely.
        try {
            String hz  = String.valueOf(targetHz);
            String hzF = targetHz + ".0";
            com.gamebooster.app.engine.PrivilegeBridgeEngine.executePrivileged(
                    "settings put system peak_refresh_rate " + hzF);
            com.gamebooster.app.engine.PrivilegeBridgeEngine.executePrivileged(
                    "settings put system min_refresh_rate " + hzF);
            com.gamebooster.app.engine.PrivilegeBridgeEngine.executePrivileged(
                    "settings put global peak_refresh_rate " + hzF);
            com.gamebooster.app.engine.PrivilegeBridgeEngine.executePrivileged(
                    "settings put global min_refresh_rate " + hzF);
            com.gamebooster.app.engine.PrivilegeBridgeEngine.executePrivileged(
                    "settings put system match_content_frame_rate 0");
            com.gamebooster.app.engine.PrivilegeBridgeEngine.executePrivileged(
                    "settings put secure match_content_frame_rate_preference 0");
            // Disable dynamic VRR / adaptive refresh so the OS can't self-throttle back to lower rates
            com.gamebooster.app.engine.PrivilegeBridgeEngine.executePrivileged(
                    "setprop persist.vendor.display.vrr.disable 1");
            com.gamebooster.app.engine.PrivilegeBridgeEngine.executePrivileged(
                    "setprop ro.surface_flinger.set_idle_timer_ms 0");
            com.gamebooster.app.engine.PrivilegeBridgeEngine.executePrivileged(
                    "setprop ro.surface_flinger.set_touch_timer_ms 0");
            anySuccess = true;
            Log.i(TAG, "Privileged shell enforced " + targetHz + "Hz settings keys");
        } catch (Throwable t) {
            Log.w(TAG, "Privileged Hz shell error: " + t.getMessage());
        }

        return anySuccess;
    }

    public static boolean forceUnlockSystemMaxRefreshRate(Context context) {
        int maxHz = Math.round(getMaxHardwareRefreshRate(context));
        int targetHz = maxHz > 0 ? maxHz : 60;
        return forceUnlockSystemRefreshRate(context, targetHz);
    }

    /**
     * Enforces highest hardware refresh rate via EVERY available path simultaneously.
     * Called on app launch, game launch, and the Hero Hardware Banner tap.
     */
    public static void forceMaxHzNeverFallback(Context context) {
        if (context == null) return;
        int maxHz = Math.round(getMaxHardwareRefreshRate(context));
        int targetHz = maxHz > 0 ? maxHz : 60;
        // Settings path
        forceUnlockSystemRefreshRate(context, targetHz);
        // Shizuku 6-layer path
        com.gamebooster.app.booster.MaxHzForceChannel.forceApply(targetHz);
    }

    /** Restores the public refresh-rate preferences this app may have set. */
    public static boolean restoreAdaptiveRefreshRate(Context context) {
        if (context == null || !Settings.System.canWrite(context)) return false;
        return Settings.System.putString(context.getContentResolver(), "min_refresh_rate", null)
                | Settings.System.putString(context.getContentResolver(), "peak_refresh_rate", null);
    }
}

File: android/app/src/main/java/com/gamebooster/app/device/HardwareDisplayController.kt
package com.gamebooster.app.device

import android.app.Activity
import android.content.Context
import android.hardware.display.DisplayManager
import android.os.Build
import android.util.Log
import android.view.Display
import android.view.Surface
import com.gamebooster.app.shizuku.ShizukuExecutor

/**
 * HardwareDisplayController — Handles display modes enumeration and forces hardware refresh rates (60Hz -> 165Hz+).
 *
 * Implements:
 *  - DisplayManager.getSupportedModes() enumeration
 *  - Activity preferredDisplayModeId injection
 *  - Surface.setFrameRate() and Surface.setFrameRateCategory()
 *  - Shizuku privileged settings overrides: peak_refresh_rate, min_refresh_rate, user_refresh_rate
 */
object HardwareDisplayController {

    private const val TAG = "HardwareDisplayCtrl"

    /**
     * Finds the maximum physical refresh rate supported by the default display hardware.
     */
    @JvmStatic
    fun getMaxHardwareRefreshRate(context: Context?): Float {
        if (context == null) return 60f
        val dm = context.getSystemService(Context.DISPLAY_SERVICE) as? DisplayManager ?: return 60f
        val display = dm.getDisplay(Display.DEFAULT_DISPLAY) ?: return 60f
        val modes = display.supportedModes ?: return 60f

        var maxRate = 60f
        for (mode in modes) {
            if (mode.refreshRate > maxRate) {
                maxRate = mode.refreshRate
            }
        }
        return maxRate
    }

    /**
     * Returns the Display.Mode with the highest refresh rate.
     */
    @JvmStatic
    fun getMaxHardwareMode(context: Context?): Display.Mode? {
        if (context == null) return null
        val dm = context.getSystemService(Context.DISPLAY_SERVICE) as? DisplayManager ?: return null
        val display = dm.getDisplay(Display.DEFAULT_DISPLAY) ?: return null
        val modes = display.supportedModes ?: return null

        var maxMode: Display.Mode? = null
        for (mode in modes) {
            if (maxMode == null || mode.refreshRate > maxMode.refreshRate) {
                maxMode = mode
            }
        }
        return maxMode
    }

    /**
     * Injects the preferred display mode into the window attributes to enforce max Hz on this Activity.
     */
    @JvmStatic
    fun applyMaxRefreshRateToWindow(activity: Activity) {
        try {
            val maxMode = getMaxHardwareMode(activity) ?: return
            val params = activity.window.attributes
            params.preferredDisplayModeId = maxMode.modeId
            activity.window.attributes = params
            Log.i(TAG, "Window preferred display mode set to modeId=${maxMode.modeId} (${maxMode.refreshRate}Hz)")
        } catch (t: Throwable) {
            Log.w(TAG, "Failed to apply preferred display mode to window: ${t.message}")
        }
    }

    /**
     * Applies frame rate hint directly to a Surface instance (Android 13+).
     */
    @JvmStatic
    fun applySurfaceFrameRate(surface: Surface, targetFps: Float) {
        if (Build.VERSION.SDK_INT >= 33) {
            try {
                surface.setFrameRate(
                    targetFps,
                    Surface.FRAME_RATE_COMPATIBILITY_FIXED_SOURCE,
                    Surface.CHANGE_FRAME_RATE_ALWAYS
                )
                Log.d(TAG, "Surface frame rate hint applied: ${targetFps}fps")
            } catch (t: Throwable) {
                Log.w(TAG, "Failed to set Surface frame rate: ${t.message}")
            }
        }

        if (Build.VERSION.SDK_INT >= 34) {
            try {
                // API 34 FRAME_RATE_CATEGORY_HIGH = 3
                val method = surface.javaClass.getMethod("setFrameRateCategory", Int::class.javaPrimitiveType)
                method.invoke(surface, 3)
                Log.d(TAG, "Surface FRAME_RATE_CATEGORY_HIGH applied via reflection")
            } catch (t: Throwable) {
                Log.d(TAG, "setFrameRateCategory not available on this platform")
            }
        }
    }

    /**
     * Force-locks the system display refresh rate to the hardware maximum using Shizuku privileged shell.
     * Prevents OEM adaptive refresh stepping down to 60Hz during gaming.
     */
    @JvmStatic
    fun forceUnlockSystemMaxRefreshRate(context: Context): Boolean {
        val maxHz = Math.round(getMaxHardwareRefreshRate(context))
        Log.i(TAG, "Hardware ceiling detected: ${maxHz}Hz. Pushing system overrides via Shizuku...")

        val commands = listOf(
            "settings put system peak_refresh_rate $maxHz",
            "settings put system min_refresh_rate $maxHz",
            "settings put system user_refresh_rate $maxHz",
            "settings put global low_power 0",
            "settings put global low_power_sticky 0",
            "settings put global adaptive_battery_management_enabled 0"
        )

        var allSuccess = true
        for (cmd in commands) {
            val res = ShizukuExecutor.executeShizukuCommand(cmd)
            if (res == null || res.startsWith("ERROR")) {
                allSuccess = false
                Log.w(TAG, "Command warning: $cmd ($res)")
            }
        }
        return allSuccess
    }
}
